### PR TITLE
Add envoy proxy config into pod level (workload page)

### PR DESCRIPTION
**Pull request by xeviknal**
Requires: https://github.com/kiali/kiali/pull/3400

The idea is to show a modal where you can both see different envoy proxy config (both raw and digested data).
In this first step, it shows the raw information:

Screen recording (18) (1)

### CODE REVIEW

**Comment by lucasponce** [comment]
One first comment.
I'd add a 'Cancel' buttom in the Envoy modal, it's same pattern used in the Wizards:

image

And it will help to organize when it will add additional steps in the future.

**Comment by lucasponce** [comment]
Minor, "Show Envoy details" -> "Show Envoy Details".
image

**Comment by lucasponce** [comment]
I'dd add an easy sorting in the table headers for:
image
image

(Note: I understand the scrolling but it looks weird, I wonder if using some padding in the inner table would make a difference).

**Comment by lucasponce** [comment]
From experience, "listeners" are also commonly used for debugging, so I've add them in the dropdown feature:
image

**Comment by lucasponce** [comment]
If no sidecar or exceptions, it should show the menu option disabled (otherwise it will fail).
Example: grafana workload.

image

**Comment by lucasponce** [comment]
Another thought, on errors like this:
image

I guess the modal should close and an error should be shown in the AlertManager.

**Comment by xeviknal** [comment]
Comments approached. This is ready for another round of reviews. Thanks

File: src/types/IstioObjects.ts [hunk 1]
```diff
@@ -139,6 +139,10 @@ export interface PodLogs {
   entries: LogEntry[];
 }
 
+export interface EnvoyProxy {
```
Comment (lucasponce) [inline src/types/IstioObjects.ts:142]: I'd add in the backend a model.EnvoyProxy { Configs: interfaces{} }, just to have this mapped.

I understand that any type would be necessary if we can't map the nested types here as discussed.

File: src/services/Api.ts [hunk 2]
```diff
@@ -495,6 +495,14 @@ export const getPodLogs = (
   return newRequest<PodLogs>(HTTP_VERBS.GET, urls.podLogs(namespace, name), params, {});
 };
 
+export const getPodEnvoyProxy = (namespace: string, pod: string) => {
+  return newRequest<EnvoyProxy>(HTTP_VERBS.GET, urls.podEnvoyProxy(namespace, pod), {}, {});
+};
+
+export const getPodEnvoyProxyResourceEntries = (namespace: string, pod: string, resource: string) => {
+  return newRequest<EnvoyProxy>(HTTP_VERBS.GET, urls.podEnvoyProxyResourceEntries(namespace, pod, resource), {}, {});
```
Comment (lucasponce) [inline src/services/Api.ts:503]: Perhaps I missed to review this in the backend, so both "getPodEnvoyProxy" and "getPodEnvoyProxyResourceEntries" return a generic EnvoyProxy ? I thought the second may return a model.Clusters, model.Routes, etc. Perhaps we could take a look if it's working due dynamic parsing but we are missing some type here.

Comment (xeviknal) [reply src/services/Api.ts:503]: You are right. This is could be one of `BootstrapSummary | ClusterSummary | ListenerSummary | RouteSummary` interfaces.

File: src/types/IstioObjects.ts [hunk 3]
```diff
@@ -139,6 +139,39 @@ export interface PodLogs {
   entries: LogEntry[];
 }
 
+export interface EnvoyProxyDump {
+  configs: any[];
+}
+
+export type EnvoyProxyResourceEntries = BootstrapSummary | ClusterSummary | ListenerSummary | RouteSummary;
```
Comment (lucasponce) [inline src/types/IstioObjects.ts:146]: I have a comment here.
These optional types look good on paper but in practice it's hard to check in runtime which type is coming from as there are not strong types.

So I guess if we can put a variable for any of these types and check if defined to see if the EnvoyProxyResourceEntries refers to a BootstrapSummary or a ClusterSummary.

Because this comes from an invocation

```
export const getPodEnvoyProxyResourceEntries = (namespace: string, pod: string, resource: string) => {
  return newRequest<EnvoyProxyResourceEntries>(
    HTTP_VERBS.GET,
    urls.podEnvoyProxyResourceEntries(namespace, pod, resource),
    {},
    {}
  );
};
```

So, perhaps this is missing in my backend PR review but instead to have this type pretty open:

```
type ResourceDump interface {
	Parse(dump *kubernetes.ConfigDump) error
}
```

I'd suggest to have an entry for the one of the "clusters", "routes", "boostrap", "listeners",

```
type ResourceDump interface {
         Clusters *models.Clusters,
         Routes *models.Routes,
         Bootstrap *models.Bootstrap,
         Listeners *models.Listeners,
	Parse(dump *kubernetes.ConfigDump) error
}
```
In that way, the caller would know the returning type just checking if that field is populated, not just relying on the parameter asked, I think.

File: src/components/Envoy/writers/BaseWriter.tsx [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        >
+          <TableHeader />
+          <TableBody />
+        </Table>
+      );
+    }
+  };
+}
+
+export const SummaryWriterBuilder = (resource: string, config: any) => {
```
Comment (lucasponce) [inline src/components/Envoy/writers/BaseWriter.tsx:60]: I know naming is hard and opinionated :).
Coming from Java background, I read "Writer" very confusing as normally it's used for an abstraction of streams data.

If this is just a table and we are in the UI, a suggestion, "SummaryTable", "ListenersTable" or using "*List" like "SummaryList", "ListenerLists".

I think that's the way we are naming these patterns in other components and just to follow it to not introduce a new one.

File: src/components/Envoy/EnvoyModal.tsx [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+};
+
+type EnvoyDetailState = {
+  config: any;
+  resource: string;
+  fetch: boolean;
+  pod: Pod;
+};
+
+export const EnvoyDetailsModal = ({ namespace, workload, isOpen, onClose }: EnvoyDetailModalProps) => (
```
Comment (lucasponce) [inline src/components/Envoy/EnvoyModal.tsx:61]: Perhaps you may want to reuse here same pattern we are using for the Wizards.

There is a skeleton used called "WorkloadWizard" which basically performs same steps this is doing:
- Passing the workload, namespace, and which sub-wizard you want to open.
- Taking care of common controls: Create/Cancel buttons, title, etc.

So, all EnvoyDetail class would suit inside the WorkloadWizard and leave as it is, but then we can share and modify aspect of the modal for same category (Workload, Services, etc).

File: src/components/Envoy/EnvoyModal.tsx [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+        );
+      });
+  };
+
+  render() {
+    const builder = SummaryWriterBuilder(this.state.resource, this.state.config);
+    const SummaryWriterComp = builder[0];
+    const summaryWriter = builder[1];
+    return (
+      <Stack gutter={GutterSize.sm}>
```
Comment (lucasponce) [inline src/components/Envoy/EnvoyModal.tsx:180]: Also, using the WorkloadWizard would help to not need the Stack component.
I think I remove it (or delegate it) to any of the subwizards to render it.

Perhaps you can take a look to the Service Wizards as well.

File: src/components/Envoy/EnvoyModal.tsx (hunk 5 above)
Comment (xeviknal) [reply src/components/Envoy/EnvoyModal.tsx:61]: I don't know if I fully understand this one.

This is not exactly a wizard, this is a regular modal, isn't it?. Perhaps, the modal you use in the `WorkloadWizard` could be extracted to a common component that everyone could use. 

On the other hand, I don't know why we should use a `sub-wizard` prop in this case. There is a drowpdown with multiple values, but they are not coming from the outside component neither the initial selection.

Regarding the common controls, this scenario doesn't need any "Create"/"Update", it is just informative actions you can do in this modal. There isn't any impact in the backend after finishing the task in the modal.

One thing I could do to fit more the "pattern" is not to have two components one for the modal and another for the content of the modal.

File: src/components/Envoy/EnvoyModal.tsx (hunk 6 above)
Comment (xeviknal) [reply src/components/Envoy/EnvoyModal.tsx:180]: I think we need the stack anyway since every modal content needs a grid to distribute the content.
I see that in the WorkloadWizard there is nothing in the content just a plain text.
Perhaps I missed something. I could use a clarification here.

**Comment by xeviknal** [comment]
@lucasponce I've approached some comments in the latest commits. I've also applied the changes in the backend to fit the data structure according the mentioned comments.

File: src/types/IstioObjects.ts [hunk 7]
```diff
@@ -139,6 +139,45 @@ export interface PodLogs {
   entries: LogEntry[];
 }
 
+export interface EnvoyProxyDump {
```
Comment (lucasponce) [inline src/types/IstioObjects.ts:142]: +1

File: src/components/IstioWizards/WorkloadWizardDropdown.tsx [hunk 8]
```diff
@@ -153,7 +174,17 @@ class WorkloadWizardDropdown extends React.Component<Props, State> {
...
     );
     // TODO WorkloadWizard component contains only 3scale actions but in the future we may need to bring it back
-    return <>{dropdown}</>;
+    return (
+      <>
+        {dropdown}
+        <EnvoyDetailsModal
+          namespace={this.props.namespace}
+          workload={this.props.workload}
+          show={this.state.showWizard}
```
Comment (lucasponce) [inline src/components/IstioWizards/WorkloadWizardDropdown.tsx:183]: Not a blocker, because today we have only "one" modal in the Workload Details.
But in previous Kiali version we used to have other modals for other extensions operations.
So, as the WorkloadWizardDropdown manages the state about when to show or not the modal, it's good to have a wrapper "WorkloadWizard" to manage this in a single place.

Otherwise when we brought back more modals in this section (others are on the way) we'll need to adjust this.

So basically the idea open the WorkloadWizard and inside the workload wizard modal says "show Envoy".

That's how it works in the Service Details.

Not a blocker, I suggest I'll send a followup PR with that, it wouldn't change anything relevant just a couple of refactoring about how to structure components.

**Review by lucasponce — APPROVED** [review]
LGTM

I have one comment that probably the WorkloadWizardDropdown shoudln't invoke directly the EnvoyModal but use the WorkloadWizard as parent "modal" wrapper to manage better the state when there are more "wizards" under the Workload details.

I'm ok to move forward and I'll send a followup PR with a couple of structure changes that won't affect into the feature.
